src/Symfony/Component/EventDispatcher/EventDispatcher.php

```diff
@@ -215,22 +224,46 @@ protected function doDispatch($listeners, $eventName, Event $event)
 
     /**
      * Sorts the internal list of listeners for the given event by priority.
-     *
-     * @param string $eventName The name of the event
      */
-    private function sortListeners($eventName)
+    private function sortListeners(string $eventName)
     {
         krsort($this->listeners[$eventName]);
         $this->sorted[$eventName] = array();
 
-        foreach ($this->listeners[$eventName] as $priority => $listeners) {
+        foreach ($this->listeners[$eventName] as &$listeners) {
             foreach ($listeners as $k => $listener) {
                 if (\is_array($listener) && isset($listener[0]) && $listener[0] instanceof \Closure) {
                     $listener[0] = $listener[0]();
-                    $this->listeners[$eventName][$priority][$k] = $listener;
                 }
                 $this->sorted[$eventName][] = $listener;
             }
         }
     }
+
+    /**
+     * Optimizes the internal list of listeners for the given event by priority.
+     */
+    private function optimizeListeners(string $eventName): array
+    {
+        krsort($this->listeners[$eventName]);
+        $this->optimized[$eventName] = array();
+
+        foreach ($this->listeners[$eventName] as &$listeners) {
+            foreach ($listeners as &$listener) {
+                $closure = &$this->optimized[$eventName][];
+                if (\is_array($listener) && isset($listener[0]) && $listener[0] instanceof \Closure) {
+                    $closure = static function (...$args) use (&$listener, &$closure) {
+                        if ($listener[0] instanceof \Closure) {
```

Comment:
Seems this if is unnecessary now

Reply:
It is: the value can change by reference before the closure is called.

Reply:
in that case is_array and isset are also needed?

Reply:
not really, because the only mutations that can happen are the one done in the same class, which only resolve the listener - never more.